### PR TITLE
[type] Support offset load in bit vectorized loop

### PR DESCRIPTION
**Pull request by TH3CHARLie**
Related issue = #1905

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: tests/python/test_bit_array_vectorization.py [hunk 1]
```diff
@@ -41,3 +41,49 @@ def verify():
     init()
     assign_vectorized()
     verify()
+
+
+@ti.test(require=ti.extension.quant, debug=True, cfg_optimization=False)
```
Comment (yuanming-hu) [inline tests/python/test_bit_array_vectorization.py:46]:
```suggestion
@ti.test(require=ti.extension.quant)
```

File: tests/python/test_bit_array_vectorization.py [hunk 2]
```diff
@@ -41,3 +41,49 @@ def verify():
...
+def test_offset_load():
+    ci1 = ti.type_factory_.get_custom_int_type(1, False)
```
Comment (yuanming-hu) [inline tests/python/test_bit_array_vectorization.py:48]:
```suggestion
    ci1 = ti.type_factory.custom_int(1, False)
```
Also the identifier should be `cu1`?

File: tests/python/test_bit_array_vectorization.py [hunk 3]
```diff
@@ -41,3 +41,49 @@ def verify():
...
+
+    x = ti.field(dtype=ci1)
+    y = ti.field(dtype=ci1)
+    z = ti.field(dtype=ci1)
+
+    N = 4096
+    n_blocks = 4
+    bits = 32
+    boundary_offset = 1024
```
Comment (yuanming-hu) [inline tests/python/test_bit_array_vectorization.py:57]:
```suggestion
    boundary_offset = 1024
    assert boundary_offset >= N // n_blocks
```

File: tests/python/test_bit_array_vectorization.py [hunk 4]
```diff
@@ -41,3 +41,49 @@ def verify():
...
+        for i, j in ti.ndrange((boundary_offset, N - boundary_offset),
+                               (boundary_offset, N - boundary_offset)):
+            x[i, j] = ti.random(dtype=ti.i32) % 2
+
+    @ti.kernel
+    def assign_vectorized():
+        ti.bit_vectorize(32)
+        for i, j in x:
+            y[i, j] = x[i, j + 1]
+            z[i, j] = x[i, j - 1]
```
Comment (yuanming-hu) [inline tests/python/test_bit_array_vectorization.py:78]: If we are ready, we can test `x[i - 1, j]/x[i + 1, j]` too. You can also allocate only the `y` field, and reuse the functions `assign_vectorized` and `verify` by adding template arguments `dx: ti.template(), dy: ti.template()`, and the loop bodies should be something like `y[i, j] = x[i + dx, j + dy]`.

File: taichi/transforms/bit_loop_vectorize.cpp [hunk 5]
```diff
@@ -45,6 +48,77 @@ class BitLoopVectorize : public IRVisitor {
...
+                std::make_unique<GlobalPtrStmt>(ptr->snodes, indices);
+            base_ptr->ret_type = new_ret_type;
+            base_ptr->is_bit_vectorized = true;
+            // load x[i, j](base)
+            DataType load_data_type(bit_array_physical_type);
+            auto load_base = std::make_unique<GlobalLoadStmt>(base_ptr.get());
+            load_base->ret_type = load_data_type;
+            // load x[i, j + 1](offsetted)
+            // since we are doing vectorization, the actual data should be x[i,
+            // j + 32]
```
Comment (yuanming-hu) [inline taichi/transforms/bit_loop_vectorize.cpp:72]:
```suggestion
            // j + vectorization_width]
```

**Review by yuanming-hu — APPROVED** [review]
Awesome!! Looks great. Just a few minor comments.

**Comment by TH3CHARLie** [comment]
Thanks for the review and help along the way!

**Review by yuanming-hu — APPROVED** [review]
Awesome! LGTM now. Thanks for the hard work!

**Pull request merged**